### PR TITLE
[mono][interp] optimize boxing for x == null where x is value type

### PR DESCRIPTION
**Pull request by EgorBo**
this is a quite popular pattern with generics - T ==/!= null introduces expensive "boxing" (allocation) for value types,

#### e.g. for `Test<int>`:
```csharp
bool Test<T>(T t)
{
    return t == null;
}
```
#### Was:
```
Runtime method: Program:Test<int> (int) 000000406D3F8400
Locals size 8, VT stack size: 0
Calculated stack size: 2, stated size: 8
IR_0000: ldloc.i4   0
IR_0002: box        0,0
IR_0005: ldnull
IR_0006: ceq.i8
IR_0007: ret
```
#### Now:
```
Runtime method: Program:Test<int> (int) 00000091171F8270
Locals size 8, VT stack size: 0
Calculated stack size: 1, stated size: 8
IR_0000: ldc.i4.0
IR_0001: ret
```

### CODE REVIEW

File: src/mono/mono/mini/interp/transform.c [hunk 1]
```diff
@@ -5319,6 +5319,20 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 
 				const gboolean vt = mint_type (m_class_get_byval_arg (klass)) == MINT_TYPE_VT;
 
+                /* box + ldnull + ceq/cgt_un -> 1/0 */
+                if (*(td->ip + 5) == CEE_LDNULL) {
+                    guint8 ceq = *(td->ip + 7);
+                    if (ceq == CEE_CEQ || ceq == CEE_CGT_UN) {
```
Comment (vargaz) [inline src/mono/mono/mini/interp/transform.c:5325]: Need to check that these two are in the same basic block.

**Comment by BrzVlad** [comment]
long term we should be able to track nullability of vars

File: src/mono/mono/mini/interp/transform.c [hunk 2]
```diff
@@ -5319,11 +5319,15 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 
 				const gboolean vt = mint_type (m_class_get_byval_arg (klass)) == MINT_TYPE_VT;
 
-                /* box + ldnull + ceq/cgt_un -> 1/0 */
+                /* box + ldnull + ceq/cgt_un -> true/false */
                 if (*(td->ip + 5) == CEE_LDNULL) {
                     guint8 ceq = *(td->ip + 7);
```
Comment (vargaz) [inline src/mono/mono/mini/interp/transform.c]: Need to check that ip+7 is inside the code as well.

File: src/mono/mono/mini/interp/transform.c [hunk 3]
```diff
@@ -5321,7 +5321,7 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 
 				// optimize "box + ldnull + ceq/cgt_un" to true/false
 				// e.g. `t == null` in generic code when t is a vt/primitive (but not nullable)
-				if (*(td->ip + 5) == CEE_LDNULL) {
+				if (td->sp + 7 >= td->stack && *(td->ip + 5) == CEE_LDNULL) {
 					guint8 ceq = *(td->ip + 7);
```
Comment (vargaz) [inline src/mono/mono/mini/interp/transform.c:5325]: This doesn't look good.

Comment (EgorBo) [reply src/mono/mono/mini/interp/transform.c:5325]: @vargaz any thoughts how to improve it? I am not yet familiar with interpreter internals 🙁

Comment (vargaz) [reply src/mono/mono/mini/interp/transform.c:5325]: Need to check ip+ 7 against the end of the code.

Comment (vargaz) [reply src/mono/mono/mini/interp/transform.c:5325]: I.e. td->ip - td->il_code + 7 < td->code_size

Comment (EgorBo) [reply src/mono/mono/mini/interp/transform.c:5325]: thanks!

**Comment by EgorBo** [comment]
will re-open later when I have more time to spend on learning the interpreter 🙂 
Closing to keep my amount of active PRs smaller.

**Pull request closed without merging**